### DISCUSSION TITLE
How to proxy websocket connection

### DISCUSSION THREAD

**Discussion by aguspdana** [vercel/community#643] [Help]
Hi,

My web app [Rust Explorer](https://www.rustexplorer.com), a rust playground, is built with Vite + React Router Dom. The back-end is hosted on a dedicated server and proxied via Vercel rewrite.

Currently it works over https by sending the output once the execution is finished. The problem with this setup is that we can't see the output in real-time.

So I'm trying to implement a real-time streaming output over websocket. The websocket route requires an authentication. To simplify the cookie management, I prefer to proxy all routes to the api server instead of using a different sub-domain.

Is it possible to proxy a websocket connection via Vercel?

Thanks.

**Comment by reubenjohn** [comment]
I'm getting a similar issue, I've got 2 servers: **User Browser** <--> **Next.Js server** <--> **Django server.**

Users connect to the Next.JS app. Any incoming requests to the Next.js server for /api/ai/* are routed to the Django server at /api/*:
```javascript
/** @type {import('next').NextConfig} */
module.exports = {
  rewrites: async () => {
    return [
      {
        source: '/api/ai/:path*',
        destination: `${process.env.AI_REST_API_URL}/:path*`,
      },
    ]
  },
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: 'avatars.githubusercontent.com',
        port: '',
        pathname: '**'
      }
    ]
  }
}
```

**User Browser** <--> **localhost:3000** <--> **localhost:8000**
Ony my local machine, both the HTTP connection (where_cat_is_hiding), and the websocket connection (4802052) go through fine
![local network inspection](https://github.com/vercel/community/assets/4802052/e100fa0a-edab-4fc7-bf69-2f2681919ff6)

This is the code to connect to the websocket:
![local console log](https://github.com/vercel/community/assets/4802052/6a69c47a-d2ec-4ebd-89f4-626919f51a9d)

However, on Vercel, the HTTP goes through but the websocket fails:
![vercel network inspection](https://github.com/vercel/community/assets/4802052/3d854676-e984-4d7d-a0c5-2d440b2600dd)
![vercel console log](https://github.com/vercel/community/assets/4802052/cbf004e8-28d8-42b8-af5b-57a35effacda)

I also tried directly connecting to the django server from my browser, and that works fine too (User Browser <--> Django Server):
![django console log](https://github.com/vercel/community/assets/4802052/6b91f7a8-1d14-4eec-b8ae-14f52adf9e97)
